iocore/net/SSLUtils.cc

```diff
@@ -1737,11 +1847,71 @@ ssl_error_t
 SSLAccept(SSL *ssl)
 {
   ERR_clear_error();
-  int ret = SSL_accept(ssl);
+
+  int ret       = 0;
+  int ssl_error = SSL_ERROR_NONE;
+
+#if TS_HAS_TLS_EARLY_DATA
+  if (SSL_version(ssl) != TLS1_3_VERSION) {
+    ret = SSL_accept(ssl);
+  } else {
+    SSLNetVConnection *netvc = SSLNetVCAccess(ssl);
+
+    if (SSLConfigParams::server_max_early_data > 0 && !netvc->early_data_finish) {
+      size_t nread;
+      uint8_t *buf = static_cast<uint8_t *>(ats_malloc(SSLConfigParams::server_max_early_data));
```

Comment:
We should use an iobuffer here instead of calling malloc().

Reply:
I'm not sure we can, it's the buffer to the call `int SSL_read_early_data(SSL *s, void *buf, size_t num, size_t *readbytes);`

Reply:
This `buf` looks temporally buffer, because the written data on this buffer will be copied to `early_data_buf `.
We can allocate this buffer on stack instead of heap.

Reply:
Another approach is, as Bryan pointed out, allocate IOBufferBlock, write data on underlay `char *` buffer through `IOBufferBlock::buf()` and append the blocks to `early_data_buf`. I think we could avoid `memcpy` with this approach.

Reply:
Susan and I were talking about this earlier and thought using iobuffer on this seems like a bit too much. As for allocating on the stack, the size could get really big, not quite sure that's a good idea for the stack.

Reply:
It looks a bit too much for me that overhead of `malloc`, `free` and `memcpy` :p

For the stack, I'm not sure I understand correctly, but the default stack size for each thread look 1MB. 16KB might fine.

Reply:
16KB is just the default size when enabling early data, but the user can set it to be a lot larger than that. I'll play around iobuffer a bit to see how that works.

Reply:
Correct me if I'm wrong, but I've got the impression that unless I allocate a new iobuffer with block sizes that's the same as the `server_max_early_data`, it will potentially need multiple reads to finish, and I'll have to manually add new blocks and use those to read after the previous block filled up.

Reply:
> unless I allocate a new iobuffer with block sizes that's the same as the server_max_early_data, it will potentially need multiple reads to finish

IIUC, we need to keep reading until `SSL_read_early_data` returns `SSL_READ_EARLY_DATA_FINISH`. There is no guarantee of giving `server_max_early_data` size buffer can read whole early data at first time.

 > and I'll have to manually add new blocks

Righ. To avoid calling `write`, the allocated blocks manually added to the MIOBuffer. 

> and use those to read after the previous block filled up

I'm not sure what you mean. But we don't need to fully fill up IOBufferBlock. 

I imagined like this. I might miss something. 

```
while (ret != SSL_READ_EARLY_DATA_FINISH) {
  IOBufferBlock *block = new_IOBufferBlock();
  block->alloc(BUFFER_SIZE_INDEX_16K);

  ret = SSL_read_early_data(ssl, block->buf(), index_to_buffer_size(BUFFER_SIZE_INDEX_16K), &nread);
  ...
  if (nread > 0) {
    block->fill(nread);
    etvc->early_data_buf->append_block(block);
  }
  ...
}
```

Reply:
Ah, ok, think I got you this time. I'll see what I can do.